testing: final conversion 4.19 CI to kube-proxy free

borkmann opened the pull request:
Re-enabling probe mode on 4.19.

Fixes: #11175

pchaigno committed:
bpf: fix program size issue for complex configs on 4.19

On Linux 4.19 (and any kernel with 4096 BPF instructions limit), when
running with kube-proxy-replacement=strict, IPv4, IPv6, fragment support
(and in debug mode), Cilium fails to load bpf_overlay and bpf_host with:

  Prog section '2/19' rejected: Argument list too long (7)
   - Type:         3
   - Attach Type:  0
   - Instructions: 4203 (107 over limit)
   - License:      GPL
  Error filling program arrays!

2/19 is section CILIUM_CALL_NODEPORT_REVNAT. We can reduce the number of
instructions in that section by breaking the IPv4 and IPv6 cases with
tail calls. Instead of introducing a new tail call on each path, we can
turn CILIUM_CALL_NODEPORT_REVNAT into CILIUM_CALL_IPV{4,6}_NODEPORT_REVNAT.

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno committed:
test: disable running with kube-proxy in 4.19 CI build

... and instead rely on kube-proxy's strict replacement.

Also disable tests that are not working on 4.19 due to missing kernel
features on NodePort side.

Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test: re-enable kube-proxy replacement probe mode in 4.19

Re-enable it such that we can run CI with kube-proxy-free BPF code alongside
kube-proxy.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: ensure we only xlate external ip when netns cookie unavailable

Commit b2c9b075ec92 ("bpf, sock: reduce xlation for externalTrafficPolicy=Local
to host_id") skips service translation for remote externalTrafficPolicy=Local
services (aka not being HOST_ID ipcache scope).

However, later on, we relaxed bpf_lxc to compile legacy service xlation back in
for services with externalIPs via 6426c8608fc3 ("bpf, external ip: fix service
xlation for containers"), which would then go and try to xlate such services
locally.

Issue is that here we don't differentiate between the different service types
so we would go and attempt to xlate. Given we don't populate service maps with
remote backends that traffic gets blackholed on bpf_lxc.

Instead, only xlate externalIP services at that layer and skip the rest, so we
let externalTrafficPolicy=Local dsts pass through and be xlated at the remote
LB. We /already/ have this kind of behavior when we have netns cookies since
the bpf_lxc xlation is compiled out entirely there, but not for older kernels,
thus, fix it to make both consistent.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -1241,28 +1264,29 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
-	(!defined(ENABLE_DSR) || \
-	 (defined(ENABLE_DSR) && defined(ENABLE_DSR_HYBRID)))
-	if ((ctx->mark & MARK_MAGIC_SNAT_DONE) != MARK_MAGIC_SNAT_DONE) {
-		ret = nodeport_nat_fwd(ctx, false);
```
I don't think the host firewall is the main contributor to the program size here. It's only enabled where kube-proxy is disabled, which is not the case of the 4.19 pipeline yet. `nodeport_nat_fwd()` is the culprit here: the IPv4 and IPv6 paths are not in separate tail calls, which is what 26a39a8c9963325f2080ac24555067e1d17ca84b addresses.

borkmann replied:
> I don't think the host firewall is the main contributor to the program size here. It's only enabled where kube-proxy is disabled, which is not the case of the 4.19 pipeline yet. `nodeport_nat_fwd()` is the culprit here: the IPv4 and IPv6 paths are not in separate tail calls, which is what [26a39a8](https://github.com/cilium/cilium/commit/26a39a8c9963325f2080ac24555067e1d17ca84b) addresses.

The PR here re-enables kube-proxy-replacement probing again, which is why it gets otherwise hit on 4.19. I basically was trying to figure out if I could get away with something less complex given the 1.8 release just around the corner, and this one seemed to have resolved the issue, at least when I tried locally on 4.19.57. Will see how it goes, otherwise I can also take 26a39a8 with the earlier mentioned tweaks back in.

borkmann replied:
Took 26a39a8 back in, the 4.19 tests in CI were still 9 insns over the limit in some cases. With 26a39a8 we can also bump the NAT collision retries again which is great. I left the CILIUM_CALLS_ definitions as-is. It's not ABI of some sort so we can change it anytime, but I think we need some more investigation on what we discussed wrt populating tail call map slots. I do agree that we have an issue here which needs reworking first. E.g. if we'd move both new CALL defs at the end and leave a 'gap' in the middle, then this might not be cleaned up, if we move both into the middle, then this might be disruptive wrt datapath upgrades.

borkmann committed:
cilium, cleanup: reorder bpf fs cleanup and remove attached BPF progs

This fixes a bug where during a Cilium upgrade test full connectivity is lost.
The major reason why it is happening is basically the test for ...

  if strings.Contains(bpfFilter.Name, "bpf_netdev.o") [...]

... which is buggy due to multi-dev support since we now end up with something
like bpf_netdev_eno1.o instead of bpf_netdev.o . This means that tc BPF progs
never get removed from the devices, yet we're tearing down everything else and
once we remove all files from the BPF fs then in case of tail call maps, the
kernel will purge all BPF programs inside tail call maps due to circular dependency
avoidance. However, this means that those still-attached programs go into a
drop-all mode due to missed tail calls.

Fix this in four ways: i) Match on "bpf_netdev" prefix instead so we catch old
but also new names, ii) Reorder BPF fs cleanup with tc BPF prog removal, meaning,
we first need to remove all progs from the devices and only then clean up BPF fs,
iii) When calling with --bpf-state only, then also remove progs from devices, and
iv) Last but not least, add also other tc prog names. Step ii) to iv) is necessary
since there can still be a small race window where we missed tail calls right
before the device BPF prog removal.

'K8sUpdates Tests upgrade and downgrade from a Cilium stable image to master' test
case is then fixed.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docs: add cilium status info to hostport section

Add a small paragraph for helping users to verify whether enabled or not.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test: use new image to clean up prior cilium state in upgrade test

Prior tests have been run with the latest Cilium image. Using the old one
for the initial cleanup is broken since it will not clean-up everything.

For example the 1.7 cleanup method will try to remove tc attached filters
with the name bpf_netdev.o, but for 1.8/latest they are names bpf_netdev_<dev>.o
which will then break connectivity after flushing all files from BPF fs as
tail call maps are purged which gets us into drop-all due to missed tail calls.

Run new (and old) to make sure there are no quirks from old version left behind
that the new one didn't catch.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: deprecate --disable-ipv4 agent option

We have --enable-{ipv4,ipv6}={true,false} setting. The --disable-ipv4 is
legacy and hidden for some time already. Schedule for removal.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium: bail out in native XDP mode when using multi-device setup

In XDP, we can only xmit via XDP_TX, so a multi-device setup would
otherwise be broken. Tell the user to specify the device manually
in case there were multiple devices detected.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docs: add notes on XDP on multi-device nodes

Add few notes on how to deal with nodes that have multiple devices which
are auto-detected. Also make it clear as a limitation.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: bump snat retries for older kernels

Given v4/v6 has been split off into different tail calls, it enables
us to significantly bump the collision retries on older kernels.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno commented on Documentation/spelling_wordlist.txt:
```diff
@@ -351,6 +352,7 @@ hashtables
 herokuapp
 hexData
 hoc
+homogenous
```
```suggestion
homogeneous
```

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -344,9 +344,9 @@ NodePort XDP Acceleration
 Cilium has built-in support for accelerating NodePort, LoadBalancer services and
 services with externalIPs for the case where the arriving request needs to be
 pushed back out of the node when the backend is located on a remote node. This
-ability to act as a hairpin load balancer can be handled by Cilium at the XDP
```
Could you also change the following above if you're touching this file (I wasn't fast enough to submit my review yesterday)?
```diff
- (one per eth0 and eth1 devices)
+ (one for each of devices eth0 and eth1)
```

borkmann replied:
np, done.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1803,15 +1803,16 @@ func initKubeProxyReplacementOptions() {
+		if option.Config.NodePortAcceleration != option.NodePortAccelerationDisabled &&
+			len(option.Config.Devices) != 1 {
+			log.Fatalf("Cannot set NodePort acceleration due to multi-device setup (%q). Specify --%s with a single device to enable NodePort acceleration.", option.Config.Devices, option.Devices)
+		}
```
@brb PTAL in context of the autodetection logic and how the devices list may be pre-populated here.

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -2142,14 +2142,6 @@ func (kub *Kubectl) overwriteHelmOptions(options map[string]string) error {
-	if os.Getenv("KERNEL") == "419" {
-		options = addIfNotOverwritten(options, "global.kubeProxyReplacement", "partial")
-		options = addIfNotOverwritten(options, "global.hostServices.enabled", "true")
-	}
```
Could you also pull in 6b14badba830a4c335d3e65fbf596deb7a1d2bab, so that kube-proxy is disabled and the code below runs? That way, `kube-proxy-replacement` would be `strict` and the host firewall would be enabled.

borkmann replied:
Yes, agree. One thing we need to think about next is that we won't have any kube-proxy based coverage aside from only 4.9 kernel in CI.

pchaigno replied:
Yep, we will even have some tests that won't run anywhere anymore. For example, https://github.com/cilium/cilium/blob/6beb6b3d5a48e552395bd62669d3b8ebcbb0c222/test/k8sT/DatapathConfiguration.go#L208-L214
I was planning to have a look at all tests once v1.8 is out to see if any need to be reenable in some way.

borkmann replied:
Agree, somehow it would be good to perhaps run 4.19 with one job in kube-proxy free and another with kube-proxy mode to be on safe side.

joestringer commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -370,6 +370,14 @@ modes and can be enabled as follows for ``nodePort.mode=hybrid`` in this example
+``eth0`` is the native XDP supported networking device. In that case, the device
+name ``eth0`` must be the same on all Cilium managed nodes. Similarly, the underlying
+driver for ``eth0`` must have native XDP support on all Cilium managed nodes. This
+is typically provided in homogenous cluster environments, for example.
```
nit:

```suggestion
driver for ``eth0`` must have native XDP support on all Cilium managed nodes.
```

I don't think the final sentence here actually clarifies anything. A homogenous environment is defined by having the same configuration in all nodes :-)

borkmann replied:
Agree, I'll take it out.

joestringer reviewed (approved):
Couple of minor comments, otherwise LGTM.

borkmann committed:
cilium: don't warn if nodeport dev for default route cannot be found

It is otherwise just confusing to users to figure out whether they need
to react to this or not. Only really warn if we cannot auto-derive it
from the k8s Node IP since this case would be unexpected.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
test: specify single device for XDP nodeport tests

... as otherwise it won't work on multi-dev node. Revert back to auto-detection
after XDP tests have run.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
docs: link to map sizing guide from XDP section

Make users aware that they might need to tweak their map sizes and
link to the guide which has more details on it.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
